yt_dlp/update.py

```diff
@@ -146,179 +164,288 @@ def _sha256_file(path):
+    checksum: str | None = None
 
-        if not self.target_tag:
-            self.target_tag = 'latest'
+    _has_update = True
+
+
+class Updater:
+    def __init__(self, ydl, target: str | None = None):
+        self.ydl = ydl
+        # For backwards compat, target needs to be treated as if it could be None
+        self.requested_channel, sep, self.requested_tag = (target or CHANNEL).rpartition('@')
+        # Check if requested_tag is actually the requested repo/channel
+        if not sep and ('/' in self.requested_tag or self.requested_tag in UPDATE_SOURCES):
+            self.requested_channel = self.requested_tag
+            self.requested_tag: str = None  # type: ignore (we set it later)
+        elif not self.requested_channel:
+            # User did not specify a channel, so we are requesting the default channel
+            self.requested_channel = CHANNEL.partition('@')[0]
+
+        # --update should not be treated as an exact tag request even if CHANNEL has a @tag
+        self._exact = bool(target) and target != CHANNEL
+        if not self.requested_tag:
+            # User did not specify a tag, so we request 'latest' and track that no exact tag was passed
+            self.requested_tag = 'latest'
             self._exact = False
-        elif self.target_tag != 'latest':
-            self.target_tag = f'tags/{self.target_tag}'
 
-        if '/' in self.target_channel:
-            self._target_repo = self.target_channel
-            if self.target_channel not in (CHANNEL, *UPDATE_SOURCES.values()):
+        if '/' in self.requested_channel:
+            # requested_channel is actually a repository
+            self.requested_repo = self.requested_channel
+            if not self.requested_repo.startswith('yt-dlp/') and self.requested_repo != ORIGIN:
                 self.ydl.report_warning(
                     f'You are switching to an {self.ydl._format_err("unofficial", "red")} executable '
-                    f'from {self.ydl._format_err(self._target_repo, self.ydl.Styles.EMPHASIS)}. '
+                    f'from {self.ydl._format_err(self.requested_repo, self.ydl.Styles.EMPHASIS)}. '
                     f'Run {self.ydl._format_err("at your own risk", "light red")}')
                 self._block_restart('Automatically restarting into custom builds is disabled for security reasons')
         else:
-            self._target_repo = UPDATE_SOURCES.get(self.target_channel)
-            if not self._target_repo:
+            # Check if requested_channel resolves to a known repository or else raise
+            self.requested_repo = UPDATE_SOURCES.get(self.requested_channel)
+            if not self.requested_repo:
                 self._report_error(
-                    f'Invalid update channel {self.target_channel!r} requested. '
+                    f'Invalid update channel {self.requested_channel!r} requested. '
                     f'Valid channels are {", ".join(UPDATE_SOURCES)}', True)
```

Comment:
need this for unit tests
```suggestion
                    f'Invalid update channel {self.requested_channel!r} requested. '
                    f'Valid channels are {", ".join(UPDATE_SOURCES)}', True)

        self._identifier = f'{detect_variant()} {system_identifier()}'
```